internal/service/resources/collector/utils.go

```diff
@@ -0,0 +1,225 @@
+		}
+
+		slog.Debug("object inserted", "table", object.TableName(), "uuid", uuid, "record", after)
+
+	case err == nil:
+		// Updated object instance
+		before = record
+
+		// We only need to update the fields that have actually changed so compare them to get the list of fields.
+		tags := utils.CompareObjects(*record, object, "CreatedAt")
+		if len(tags) == 0 {
+			// This shouldn't happen since the generation id always changes
+			after = before
+			slog.Warn("no change detected on persisted object", "table", object.TableName(), "uuid", uuid)
+			return before, after, nil
+		}
+
+		after, err = utils.Update[T](ctx, db, object, uuid, tags.Fields()...)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to update object '%s/%s': %w", object.TableName(), uuid, err)
+		}
+
+		slog.Debug("object updated",
+			"table", object.TableName(), "uuid", uuid, "before", before, "after", after, "columns", tags.Fields())
+	default:
+		return nil, nil, fmt.Errorf("failed to get object '%s/%s': %w", object.TableName(), uuid, err)
+	}
+
+	return before, after, nil
+}
+
+// persistDataChangeEvent persists a data change object to its database table.  The before and
+// after model objects are marshaled to JSON prior to being stored.
+func persistDataChangeEvent(ctx context.Context, db *pgxpool.Pool, tableName string, uuid uuid.UUID,
+	parentUUID *uuid.UUID, before, after any) (*models.DataChangeEvent, error) {
+	var err error
+	var beforeJSON, afterJSON []byte
+	if before != nil {
+		if beforeJSON, err = json.Marshal(before); err != nil {
+			return nil, fmt.Errorf("failed to marshal before object: %w", err)
+		}
+	}
+	if after != nil {
+		if afterJSON, err = json.Marshal(after); err != nil {
+			return nil, fmt.Errorf("failed to marshal after object: %w", err)
+		}
+	}
+
+	dataChangeEvent := models.DataChangeEvent{
+		ObjectType: tableName,
+		ObjectID:   uuid,
+		ParentID:   parentUUID,
+	}
+
+	if beforeJSON != nil {
+		value := string(beforeJSON)
+		dataChangeEvent.BeforeState = &value
+	}
+	if afterJSON != nil {
+		value := string(afterJSON)
+		dataChangeEvent.AfterState = &value
+	}
+
+	result, err := utils.Create[models.DataChangeEvent](ctx, db, dataChangeEvent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create data change event: %w", err)
+	}
+
+	return result, nil
+}
+
+// persistObjectWithChangeEvent persists an object to its database table and if the external API
+// model representation of the object has changed then a data change event is stored.  Persisting
+// of the object and its change event are captured under the same transaction to ensure we never
+// lose any change events.
+func persistObjectWithChangeEvent[T db.Model](ctx context.Context, db *pgxpool.Pool, record T,
+	uuid uuid.UUID, parentUUID *uuid.UUID,
+	converter GenericModelConverter) (*models.DataChangeEvent, error) {
+	var dataChangeEvent *models.DataChangeEvent
+	txCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	tx, err := db.Begin(txCtx)
```

Comment:
this seems not to be ok. You are not doing anything with the transaction structure (tx).

the defer tx.Rollback() is also missing. For queries to be part of the transaction, they should be called from methods of the tx structure. https://pkg.go.dev/github.com/jackc/pgx#Tx There are Exec and QueryRow

Reply:
I'm not sure that's true.  The `BeginTx` just stores the current `conn` into the `Tx` structure and those delegated methods on `Tx` just call the underlying connection.  I'll take a closer look tomorrow as I missed this comment when I fixed the others from today.  

As for the tx.Rollback() it is not necessary if you do a `defer cancel()` on the context passed to `Begin()`... it does nothing if the transaction was already commited, and rolls it back otherwise.

Reply:
Also, when I tested this it seemed to be working as expected... but i'll double-check.

Reply:
That is not true. We are using a pool here. `Begin` calls `BeginTx` which acquires a connection and then stores that connection into Tx. https://github.com/jackc/pgx/blob/v5.5.5/pgxpool/pool.go#L653

The transaction is relevant for that connection. From our utils/repository, we call `Query` which acquires another connection.

And Rollback is needed, the context only applies for the Begin method. See comment

```
// Begin acquires a connection from the Pool and starts a transaction. Unlike database/sql, the context only affects the begin command. i.e. there is no
// auto-rollback on context cancellation. Begin initiates a transaction block without explicitly setting a transaction mode for the block (see BeginTx with TxOptions if transaction mode is required).
// *pgxpool.Tx is returned, which implements the pgx.Tx interface.
// Commit or Rollback must be called on the returned transaction to finalize the transaction block.
```

Reply:
thanks, I'll rework this a bit

Reply:
Fixed